daemons/based/pacemaker-based.c

```diff
@@ -52,159 +55,190 @@ static bool startCib(const char *filename);
+ *
+ * \param[out] error  GLib error object
+ *
+ * \return Standard Pacemaker return code
+ */
+static int
+setup_stand_alone(GError **error)
+{
+    int rc = 0;
+    struct passwd *pwentry = NULL;
+
+    preserve_status = TRUE;
+    cib_writes_enabled = FALSE;
+
+    errno = 0;
+    pwentry = getpwnam(CRM_DAEMON_USER);
+    if (pwentry == NULL) {
+        exit_code = CRM_EX_FATAL;
+        if (errno != 0) {
+            g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                        "Error getting password DB entry for %s: %s",
+                        CRM_DAEMON_USER, strerror(errno));
+            return errno;
+        }
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Password DB entry for '%s' not found", CRM_DAEMON_USER);
+        return ENXIO;
+    }
+
+    rc = setgid(pwentry->pw_gid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not set group to %d: %s",
+                    pwentry->pw_gid, strerror(errno));
+        return errno;
+    }
+
+    rc = initgroups(CRM_DAEMON_USER, pwentry->pw_gid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not setup groups for user %d: %s",
+                    pwentry->pw_uid, strerror(errno));
+        return errno;
+    }
+
+    rc = setuid(pwentry->pw_uid);
+    if (rc < 0) {
+        exit_code = CRM_EX_FATAL;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "Could not set user to %d: %s",
+                    pwentry->pw_uid, strerror(errno));
+        return errno;
+    }
+    return pcmk_rc_ok;
+}
+
+static GOptionEntry entries[] = {
+    { "stand-alone", 's', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &stand_alone,
+      "(Advanced use only) Run in stand-alone mode", NULL },
+
+    { "disk-writes", 'w', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE,
+      &cib_writes_enabled,
+      "(Advanced use only) Enable disk writes (enabled by default unless in "
+      "stand-alone mode)", NULL },
+
+    { "cib-root", 'r', G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING, &cib_root,
+      "(Advanced use only) Directory where the CIB XML file should be located "
+      "(default: " CRM_CONFIG_DIR ")", NULL },
+
+    { NULL }
 };
 
+static GOptionContext *
+build_arg_context(pcmk__common_args_t *args)
+{
+    GOptionContext *context = NULL;
+
+    context = pcmk__build_arg_context(args, NULL, NULL, "[metadata]");
+    pcmk__add_main_args(context, entries);
+    return context;
+}
+
 int
 main(int argc, char **argv)
 {
-    int flag;
-    int rc = 0;
-    int index = 0;
-    int argerr = 0;
-    struct passwd *pwentry = NULL;
+    int rc = pcmk_rc_ok;
     crm_ipc_t *old_instance = NULL;
 
+    GError *error = NULL;
+
+    pcmk__common_args_t *args = pcmk__new_common_args(SUMMARY);
+    gchar **processed_args = pcmk__cmdline_preproc(argv, NULL);
```

Comment:
Any short options that take an argument need to be added to the second parameter to pcmk__cmdline_preproc.